### PR TITLE
Clarify that time marches on may run at any rate

### PR DESCRIPTION
**Pull request by tidoust**
The text suggested that HTML requires the time marches on steps and timeupdate events to run and get fired every 250 milliseconds at a minimum. I believe that is incorrect.

HTML does not require a specific rate to run the time marches on steps. The wording used in the time marches on step that talks about `timeupdate` events assumes that the steps run at least 4 times per second, but that is an implicit assumption which is not required anywhere AFAICT. The 250ms upper bound is a suggestion, not a requirement (and the informative note that follows the step is confusing because user agents can fire events at a slower rate in practice).

The only requirements in HTML are:
- "When the current playback position of a media element changes (e.g. due to playback or seeking), the user agent must run the time marches on steps" (see [requirement](https://html.spec.whatwg.org/multipage/media.html#playing-the-media-resource:current-playback-position-13)), which essentially requires the rate at which the time marches on steps run to match the rate at which the current playback position changes; and
- "When a media element is potentially playing [...], its current playback position must increase monotonically at the element's playbackRate units of media time per unit time of the media timeline's clock" (see [requirement](https://html.spec.whatwg.org/multipage/media.html#playing-the-media-resource:current-playback-position-9)), which sets the speed at which the current playback position increases, but not the rate at which it gets updated.

New text clarifies that the issue is that implementations are only required to run the time marches on steps when the current playback position changes, and that they are basically free to do that at any rate. On top of that, the time marches on steps allow timeupdate events to be fired only every 250ms, which makes the mechanism unsuitable for synchronized rendering. The HTML spec explicitly [discourages the use of `timeupdate` events for processing cues](https://html.spec.whatwg.org/multipage/media.html#best-practices-for-metadata-text-tracks:event-media-timeupdate)

***
<a href="https://pr-preview.s3.amazonaws.com/tidoust/me-media-timed-events/pull/35.html" title="Last updated on Feb 16, 2019, 1:02 PM UTC (607602a)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/me-media-timed-events/35/f5a1428...tidoust:607602a.html" title="Last updated on Feb 16, 2019, 1:02 PM UTC (607602a)">Diff</a>

### CODE REVIEW

**Comment by chrisn** [comment]
Step 6 in time marches on seems normative to me (`timeupdate` to happen after at most 250 ms + time to run the previous `timeupdate` handler). The only way `timeupdate` can be fired during normal playback is by running time marches on. Therefore there's an implied requirement to run time marches on at that rate.

**Comment by tidoust** [comment]
Step 6 is normative but I think we're reading the step differently. Where does it say that `timeupdate` is to happen after at most 250ms? I read the step as asking the user agent to choose a time between 15 and 250ms and to check, when this step is run, whether more than this value has elapsed since last time such an event was fired. The value selected by the user agent (e.g. 250ms) is a lower bound of a desired interval, not an upper bound.

In any case, what I'm proposing is actually not to read too much into implicit assumptions and requirements, not to mention `timeupdate` events when it's not needed, and to point out that HTML does not set the rate at which the time marches on steps need to run explicitly at a value that would allow frame accuracy.

**Comment by nigelmegitt** [comment]
My reading matches @chrisn 's and seems to be backed up by the note under step 6:

> The event thus is not to be fired faster than about 66Hz or slower than 4Hz (assuming the event handlers don't take longer than 250ms to run).

And it is clear that it is possible that even if the event is _cued_ every 250ms the handlers themselves may be _called_ even less frequently depending on what other scripts are running.

Having said that, I don't see a problem with your proposed changes @tidoust which clearly tie the issue to real world problems in existing implementations without suggesting those implementations are not spec-compliant.

**Comment by chrisn** [comment]
Re-reading step 6:

> if the user agent has not fired a `timeupdate` event at the element in the past 15 to 250ms and is not still running event handlers for such an event, then the user agent must queue a task to fire an event ...

What should the behaviour be if the UA _is_ "still running event handlers for such an event"? When the next `timeupdate` event should be fired seems undefined in this case. Otherwise it would be every 250ms.

File: index.html [hunk 1]
```diff
@@ -624,13 +629,13 @@ <h3>Synchronization of text track cue rendering</h3>
         </p>
         <p>
           The <a>time marches on</a> steps in [[HTML]] control the firing of cue
-          events during media playback. <a>Time marches on</a> requires a
-          <code>timeupdate</code> event to be fired at the
```
Comment (nigelmegitt) [inline index.html]: How about rewording as:

```
<a>Time marches on</a> requires the <a>timeupdate</a> event to be fired
from time to time during media playback, and step 6 of that algorithm includes the note: 
"The event thus is not to be fired faster than about 66Hz or slower than 4Hz
(assuming the event handlers don't take longer than 250ms to run)."  In practice it
...
```

and then continue with this pull requests content from line 636 onwards?

**Comment by tidoust** [comment]
That's fine, let's assume your reading is correct and let's drop this. The conclusion is the same in any case: the rate at which the timing marches on steps need to run should match the framerate of the video.

**Comment by tidoust** [comment]
I'm closing this without merging following discussion in this thread and during the Media Timed Events Task Force call yesterday.
